### DISCUSSION TITLE
Question - refactoring to plug-ins

### DISCUSSION THREAD

**Discussion by alexnaspo** [haskell/haskell-language-server#2145] [Ideas]
Are there any pieces of functionality that are currently implemented in ghcide (or anywhere else) that should be moved to HLS plugins?

If so, I would love to contribute to any of these initiatives.

**Comment by jneira** [comment]
Hi, many thanks for willing to contribute to the repo.
The first one of pending refactors which come to my mind is the move of hlint tests to its plugin. Now they live in the main test component of hls:

https://github.com/haskell/haskell-language-server/blob/a800b9d71e4eac70d672d30b2488949ea26dcbc3/test/functional/FunctionalCodeAction.hs#L40

But it should be in the plugin package:

https://github.com/haskell/haskell-language-server/tree/master/plugins/hls-hlint-plugin

**Reply by alexnaspo** [reply]
I'll give this a shot

**Comment by alexnaspo** [comment]
What about moving the import completions into it's own plugin? I think I see a nice way of doing this. @jneira

**Reply by jneira** [reply]
sure, it sounds great but i would like to know the thoughts about from @pepeiborra

**Reply by pepeiborra** [reply]
I'm not particularly keen on this. The reason is that every completions plugin will incur an overhead in building the context and matching on it, leading to duplicate work as well as duplicate logic. In my opinion there are other areas of contribution with much higher impact.
